src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectAsynchronousTasksServiceExtensions.cs

```diff
@@ -0,0 +1,45 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem
+{
+    /// <summary>
+    ///     Provides extension methods for <see cref="IProjectAsynchronousTasksService"/> instances.
+    /// </summary>
+    internal static class ProjectAsynchronousTasksServiceExtensions
+    {
+        /// <summary>
+        ///     Provides protection for the specified action that the project will not close before it has completed.
+        /// </summary>
+        /// <exception cref="ArgumentNullException">
+        ///     <paramref name="asyncTaskService"/> is <see langword="null"/>.
+        ///     <para>
+        ///         -or-
+        ///     </para>
+        ///     <paramref name="action"/> is <see langword="null"/>.
+        /// </exception>
+        /// <exception cref="OperationCanceledException">
+        ///     Thrown if the project was already unloaded before this method was invoked.
+        /// </exception>
+        /// <remarks>
+        ///     This method differs from the <see cref="CommonProjectSystemTools.LoadedProjectAsync(IProjectAsynchronousTasksService, Func{Task}, bool)"/> 
+        ///     method, in that if the project is in the process of being unloaded by the time the action is executed, then it bails earlier by throwing 
+        ///     <see cref="OperationCanceledException"/>.
+        /// </remarks>
+        public static JoinableTask LoadedProjectAvoidingUnnecessaryWorkAsync(this IProjectAsynchronousTasksService asyncTaskService, Func<Task> action)
+        {
+            Requires.NotNull(asyncTaskService, nameof(asyncTaskService));
+            Requires.NotNull(action, nameof(action));
+
+            return asyncTaskService.LoadedProjectAsync(() => {
+
+                asyncTaskService.UnloadCancellationToken.ThrowIfCancellationRequested();
```

Comment:
Help me understand this stack https://github.com/dotnet/project-system/issues/2377:

```
       Microsoft.VisualStudio.ProjectSystem.Managed.dll!Microsoft.VisualStudio.ProjectSystem.LanguageServices.LanguageServiceHost.ExecuteWithinLockAsync.AnonymousMethod__0() Line 159	C#
 	[Async Call]	
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.JoinAsync(System.Threading.CancellationToken cancellationToken) Line 519	C#
 	[Async Call]	
 	Microsoft.VisualStudio.ProjectSystem.Managed.dll!Microsoft.VisualStudio.ProjectSystem.LanguageServices.LanguageServiceHost.HandleAsync(Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.IProjectSubscriptionUpdate> update, Microsoft.VisualStudio.ProjectSystem.LanguageServices.RuleHandlerType handlerType) Line 286	C#
 	[Async Call]	
 	Microsoft.VisualStudio.ProjectSystem.Managed.dll!Microsoft.VisualStudio.ProjectSystem.LanguageServices.LanguageServiceHost.OnProjectChangedCoreAsync.AnonymousMethod__0() Line 112	C#
 	[Async Call]	
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.JoinAsync(System.Threading.CancellationToken cancellationToken) Line 519	C#
 	Microsoft.VisualStudio.ProjectSystem.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.ProjectAsynchronousTasksServiceBase.DrainTaskQueueAsync(Microsoft.VisualStudio.ProjectSystem.ProjectCriticalOperation operation, bool drainCurrentQueueOnly, bool throwExceptions, System.Threading.CancellationToken cancellationToken) Line 288	C#
 	[Async Call]	
 	Microsoft.VisualStudio.ProjectSystem.Managed.dll!Microsoft.VisualStudio.ProjectSystem.LanguageServices.LanguageServiceHost.OnProjectChangedCoreAsync(Microsoft.VisualStudio.ProjectSystem.IProjectVersionedValue<Microsoft.VisualStudio.ProjectSystem.IProjectSubscriptionUpdate> e, Microsoft.VisualStudio.ProjectSystem.LanguageServices.RuleHandlerType handlerType) Line 110	C#
 	[Async Call]	
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.Close.AnonymousMethod__441_0() Line 1236	C#
```

It appears as if Close itself is causing the task to run, is that not the case?